kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java

```diff
@@ -62,30 +68,33 @@ public InBandKeyManager(Kms<K, E> kms,
         this.dekTtlNanos = 5_000_000_000L;
         this.maxEncryptionsPerDek = maxEncryptionsPerDek;
         // TODO This ^^ must be > the maximum size of a batch to avoid an infinite loop
-        this.keyContextCache = new ConcurrentHashMap<>();
+        this.keyContextCache = Caffeine.newBuilder()
+                .evictionListener(this::onRemoval)
+                .removalListener(this::onRemoval)
+                .buildAsync((key, executor) -> makeKeyContext(key));
         this.decryptorCache = new ConcurrentHashMap<>();
     }
 
-    private CompletionStage<KeyContext> getKeyContext(K key,
-                                                      Supplier<CompletionStage<KeyContext>> valueSupplier) {
-        return keyContextCache.compute(key, (k, v) -> {
-            if (v == null) {
-                return valueSupplier.get();
-                // TODO what happens if the CS complete exceptionally
-                // TODO what happens if the CS doesn't complete at all in a reasonably time frame?
-            }
-            else {
-                return v;
-            }
-        });
+    private void onRemoval(K key, KeyContext value, RemovalCause cause) {
+        if (value != null) {
+            value.destroy();
+        }
     }
 
     private CompletionStage<KeyContext> currentDekContext(@NonNull K kekId) {
-        return getKeyContext(kekId, makeKeyContext(kekId));
+        // todo should we add some scheduled timeout as well? or should we rely on the KMS to timeout appropriately.
+        return keyContextCache.get(kekId);
     }
 
-    private Supplier<CompletionStage<KeyContext>> makeKeyContext(@NonNull K kekId) {
-        return () -> kms.generateDekPair(kekId)
+    private CompletableFuture<KeyContext> makeKeyContext(@NonNull K kekId) {
+        return attemptMakeDekContext(kekId, 0);
+    }
+
+    private CompletableFuture<KeyContext> attemptMakeDekContext(@NonNull K kekId, int attempt) {
+        if (attempt == MAX_ATTEMPTS) {
+            return CompletableFuture.failedFuture(new DekEncryptorCreationException("failed to create DEK encryptor after " + attempt + " attempts"));
```

Comment:
"DEK encryptor" is a bit ambiguous. It could be taken to mean something that encrypts DEKs. Can we just call it an encryptor?